Adding support for ProxySQL

bmildren opened the pull request:
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

database/proxysql/proxysql_backend_servers.py
database/proxysql/proxysql_global_variables.py
database/proxysql/proxysql_manage_config.py
database/proxysql/proxysql_mysql_users.py
database/proxysql/proxysql_query_rules.py
database/proxysql/proxysql_replication_hostgroups.py
database/proxysql/proxysql_scheduler.py
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### SUMMARY

```
This PR adds support for proxysql, a popular proxy solution when using MySQL.   More detail about proxysql can be found here: https://github.com/sysown/proxysql.
```

This was previously submitted and merged to ansible-modules-extras (https://github.com/ansible/ansible-modules-extras/pull/2917), but I missed it had to be rolled back due to a couple of rogue non-ascii chars causing issues with ansible-doc.

@abadger To confirm I've run ansible-doc against each of the modules and it now runs without error 👍

Commit:
Adding support for ProxySQL

alikins commented:
Looks like the use of iteritems is failing some of the static analysis.
```
2017-01-04 16:00:43 iteritems has been removed in python3.  Alternatives:
2017-01-04 16:00:43     for KEY, VALUE in DICT.items():
2017-01-04 16:00:43     from ansible.module_utils.six import iteritems ; for KEY, VALUE in iteritems(DICT):
2017-01-04 16:00:43     from ansible.compat.six import iteritems ; for KEY, VALUE in iteritems(DICT):
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_query_rules.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_query_rules.py:            for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_query_rules.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_query_rules.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_query_rules.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_scheduler.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_scheduler.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_scheduler.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_scheduler.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_backend_servers.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_backend_servers.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_backend_servers.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_mysql_users.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_mysql_users.py:        for col, val in self.config_data.iteritems():
2017-01-04 16:00:43 ./lib/ansible/modules/database/proxysql/proxysql_mysql_users.py:        for col, val in self.config_data.iteritems():
```

bmildren commented:
Indeed it does :)  This must be an additional test that wasn't included in the ansible-modules-extras repo.. ..I'll get it updated 👍

jctanner commented:
These modules were once merged and then reverted: https://github.com/ansible/ansible-modules-extras/pull/2917

mattclay commented:
Yes, these tests are new.

For reference, the full logs of the above mentioned errors are available here: https://app.shippable.com/runs/589240b18b639e1000268148/1/console

Commit:
imported iteritems from ansible.module_utils.six where script contained reference to iteritems

Commit:
updated use of iteritems()

Commit:
updated a few under indented lines in one of the scripts to appease the pep8 god

Commit:
added ANSIBLE_METADATA to each module

Commit:
updated ANSIBLE_METADATA.status to be a list

bmildren commented:
Updated the modules to import the iteritems from ansible.module_utils.six, cleaned up a few under indented lines in one of the modules, and added ANSIBLE_METADATA to each module.

bmildren commented:
Just to confirm - does this still need additional community reviews?  Or are the community reviews from the previous merge to the extras repo sufficient here: https://github.com/ansible/ansible-modules-extras/pull/2917 ? (It was previously reverted due to a couple of issues with ansible docs, however these have now been resolved)

renecannao commented:
Please 🐑 it!

mysqlboy commented:
shipit

roman-vynar commented:
🐑 it!

akshaysuryawanshi commented:
shipit

robbwagoner commented:
🐑 it

gundalow commented:
FYI it's "shipit" for the Bot to do it's thing

gundalow commented:
bot_status

gundalow commented:
The code & docs could be simplified a bit by 
* Shared documentation (for example describing common arguments) can be placed in lib/ansible/utils/module_docs_fragments/.
* Shared code can be placed into lib/ansible/module_utils/

http://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html

gundalow commented on lib/ansible/modules/database/proxysql/proxysql_backend_servers.py:
```diff
@@ -0,0 +1,554 @@
+        is 1000.
+  max_replication_lag:
+    description:
+      - If greater than 0, ProxySQL will reguarly monitor replication lag. If
```
reguarly ->regularly

gundalow commented:
needs_revision

Please consider moving shared code out

As stated on http://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html#your-first-pull-request it's one PR per file

abadger commented:
Merged and cherrypicked to stable-2.3.

abadger commented:
We discussed this for a bit and decided that we'd merge this since it had been previously approved and we dropped the ball on re-merging it once the docs problems were addressed.  gundalow's comments are good things to address in a followup PR, though.  They would make the code more maintainable.
